### PR TITLE
Better error handling

### PR DESCRIPTION
**Pull request by zelima**
- custom error handler replaced with new exception class following the flask exception handling structure http://flask.pocoo.org/docs/0.12/patterns/apierrors/ 
- handlers for 400, 401, 404 and 500 registered at top level with 
- controllers refactored to handle internal server errors without try/catch
- new tests, as previous ones were not full covering all cases
- tests refactored to assert new error messages
- handling errors in controllers if something goes wrong with bitstore

### CODE REVIEW

File: tests/auth/test_controller.py [hunk 1]
```diff
@@ -47,7 +47,7 @@ def test_throw_400_if_user_name_and_email_is_none(self):
                               content_type='application/json')
         data = json.loads(rv.data)
         assert rv.status_code == 400
-        assert data['error_code'] == 'INVALID_INPUT'
+        assert data['message'] == 'User name or email both can not be empty'
```
Comment (anuveyatsu) [inline tests/auth/test_controller.py]: The message is not clear here.. why has it "both" word? Probably it should be:
Username and email can not be empty.

File: app/search/models.py [hunk 2]
```diff
@@ -29,7 +29,7 @@ def _build_sql_query(self, query, query_filters):
             if filter_class == 'publisher':
                 sa_filters.append(Publisher.name == filter_term)
             else:
-                raise Exception("not supported any other filter right now")
+                raise InvalidUsage("not supported any other filter right now")
```
Comment (anuveyatsu) [inline app/search/models.py]: Probably:
"any other filters are not supported right now"

File: app/auth/controllers.py [hunk 3]
```diff
@@ -122,54 +115,40 @@ def get_jwt():
...
-        app.logger.error(e)
-        return handle_error('GENERIC_ERROR', e.message, 500)
+    data = request.get_json()
+    user_name = data.get('username', None)
+    email = data.get('email', None)
+    secret = data.get('secret', None)
+    verify = False
+    user_id = None
+    if user_name is None and email is None:
+        raise InvalidUsage('User name or email both can not be empty', 400)
```
Comment (anuveyatsu) [inline app/auth/controllers.py]: The message is not clear here.. why has it "both" word? Probably it should be:
Username and email can not be empty.

Comment (rufuspollock) [inline app/auth/controllers.py]: I think it is "User name and email cannot both be empty". Note I'm not even clear where this error message comes from - when do we submit both of these?

**Pull request merged**